legistar/bills.py

```diff
@@ -414,17 +414,40 @@ def _version_rank(self, version):
 
     def relations(self, matter_id):
         relations = self.endpoint('/matters/{0}/relations', matter_id)
-        if relations:
-            highest_flag = max(int(relation['MatterRelationFlag'])
-                               for relation in relations)
 
-            relations = [relation for relation in relations
-                         if relation['MatterRelationFlag'] == highest_flag]
+        if relations:
+            return self._filter_relations(relations)
 
-            return relations
         else:
             return []
 
+    def _filter_relations(self, relations):
+        '''
+        Sometimes, many versions of a bill are related. This method returns the
+        most recent version of each relation. Override this method to apply a
+        different filter or return the full array of relations.
+        '''
+        # Sort relations such that the latest version of each matter
+        # ID is returned first.
+        sorted_relations = sorted(
+            relations,
+            key=lambda x: (
+                x['MatterRelationMatterId'],
+                x['MatterRelationFlag']
+            ),
+            reverse=True
+        )
+
+        # Assumes that there will not be more than 10 versions of a relation.
+        seen_relations = deque([], maxlen=10)
```

Comment:
i think a set object would be safer and clearer here. is there a reason to strongly prefer a deque?

Reply:
`relations` is a list of dictionaries. Dictionaries are not hashable, so `relations` can't be cast to a set. Also, relation objects referring to the same related bill are not identical duplicates:

```xml
<GranicusMatterRelation>
<MatterRelationFlag>1</MatterRelationFlag>
<MatterRelationGuid>1B541337-1A65-4D93-A6B3-BCFE4F5DA75C</MatterRelationGuid>
<MatterRelationId>1195</MatterRelationId>
<MatterRelationLastModifiedUtc>2017-09-20T22:38:45.3533333</MatterRelationLastModifiedUtc>
<MatterRelationMatterId>4333</MatterRelationMatterId>
<MatterRelationRowVersion>AAAAAACJaxw=</MatterRelationRowVersion>
</GranicusMatterRelation>

<GranicusMatterRelation>
<MatterRelationFlag>2</MatterRelationFlag>
<MatterRelationGuid>7863769D-6A6B-46AB-8F86-47CD6D0E7C33</MatterRelationGuid>
<MatterRelationId>1206</MatterRelationId>
<MatterRelationLastModifiedUtc>2017-10-09T22:27:14.51</MatterRelationLastModifiedUtc>
<MatterRelationMatterId>4333</MatterRelationMatterId>
<MatterRelationRowVersion>AAAAAACOQLs=</MatterRelationRowVersion>
</GranicusMatterRelation>
```

Reply:
that's true, but you are just checking relation_id which is hashable?

Reply:
Oh, I see. I misunderstood your suggestion. Done.